presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSourceFactory.java

```
@@ -93,7 +88,25 @@ public PageFilePageSourceFactory(
             throw new PrestoException(HIVE_CANNOT_OPEN_SPLIT, splitError(e, path, start, length), e);
         }
 
-        return Optional.of(pageFilePageSource);
+        try {
+            PageFilePageSource pageFilePageSource = new PageFilePageSource(inputStream, start, length, fileSize, blockEncodingSerde, columns);
+            return Optional.of(pageFilePageSource);
+        }
+        catch (Throwable e) {
+            try {
+                inputStream.close();
+            }
+            catch (IOException ignored) {
+            }
+            if (e instanceof PrestoException) {
+                throw (PrestoException) e;
+            }
+            String message = splitError(e, path, start, length);
+            if (e instanceof PageFileCorruptionException) {
+                throw new PrestoException(HIVE_BAD_DATA, message, e);
```

Why not to throw this exception directly?